dynatrace/stackstate_checks/dynatrace/dynatrace.py

```diff
@@ -0,0 +1,439 @@
+    "service": "api/v1/entity/services"
+}
+
+
+class DynatraceEvent(Model):
+    eventId = IntType()
+    startTime = IntType()
+    endTime = IntType()
+    entityId = StringType()
+    entityName = StringType()
+    severityLevel = StringType()
+    impactLevel = StringType()
+    eventType = StringType()
+    eventStatus = StringType()
+    tags = ListType(DictType(StringType))
+    id = StringType()
+    source = StringType()
+
+
+class State(Model):
+    last_processed_event_timestamp = IntType(required=True)
+
+
+class InstanceInfo(Model):
+    url = URLType(required=True)
+    token = StringType(required=True)
+    instance_tags = ListType(StringType, default=[])
+    events_boostrap_days = IntType(default=EVENTS_BOOSTRAP_DAYS_DEFAULT)
+    events_process_limit = IntType(default=EVENTS_PROCESS_LIMIT_DEFAULT)
+    verify = BooleanType(default=VERIFY_HTTPS)
+    cert = StringType()
+    keyfile = StringType()
+    timeout = IntType(default=TIMEOUT_DEFAULT)
+    domain = StringType(default='dynatrace')
+    environment = StringType(default='production')
+    state = ModelType(State)
+
+
+class DynatraceCheck(AgentCheck):
+    INSTANCE_TYPE = "dynatrace"
+    SERVICE_CHECK_NAME = "dynatrace"
+    INSTANCE_SCHEMA = InstanceInfo
+
+    def get_instance_key(self, instance_info):
+        return StackPackInstance(self.INSTANCE_TYPE, str(instance_info.url))
+
+    def check(self, instance_info):
+        try:
+            if not instance_info.state:
+                # Create state on the first run
+                empty_state_timestamp = self._generate_bootstrap_timestamp(instance_info.events_boostrap_days)
+                self.log.debug('Creating new empty state with timestamp: %s', empty_state_timestamp)
+                instance_info.state = State({'last_processed_event_timestamp': empty_state_timestamp})
+            self.start_snapshot()
+            self._process_topology(instance_info)
+            self._process_events(instance_info)
+            self.stop_snapshot()
+            msg = "Dynatrace check processed successfully"
+            self.service_check(self.SERVICE_CHECK_NAME, AgentCheck.OK, tags=instance_info.instance_tags, message=msg)
+        except EventLimitReachedException as e:
+            self.log.exception(str(e))
+            self.service_check(self.SERVICE_CHECK_NAME, AgentCheck.WARNING, tags=instance_info.instance_tags,
+                               message=str(e))
+        except Exception as e:
+            self.log.exception(str(e))
+            self.service_check(self.SERVICE_CHECK_NAME, AgentCheck.CRITICAL, tags=instance_info.instance_tags,
+                               message=str(e))
+
+    def _process_topology(self, instance_info):
+        """
+        Collects components and relations for each component type from dynatrace smartscape topology API
+        """
+        start_time = datetime.now()
+        self.log.info("Starting the collection of topology")
```

Comment:
I still remember leaving all log level as `INFO`, I think it should move to `DEBUG`.

Reply:
Moved log level to debug